state-sync/state-sync-v2/state-sync-driver/src/driver.rs

```diff
@@ -444,15 +449,38 @@ impl<
 
     /// Checks if the node has successfully reached the sync target
     async fn check_sync_request_progress(&mut self) -> Result<(), Error> {
-        if !self.consensus_notification_handler.active_sync_request() {
+        if !self.active_sync_request() {
             return Ok(());
         }
 
+        // There's an active sync request. Before checking if we've hit the target,
+        // wait for the storage synchronizer to drain first (to avoid preemptively
+        // notifying consensus).
+        while self.storage_synchronizer.pending_storage_data() {
+            sample!(
+                SampleRate::Duration(Duration::from_secs(PENDING_DATA_LOG_FREQ_SECS)),
+                info!("Waiting for the storage synchronizer to handle pending data!")
+            );
+        }
+
+        // Check if we've hit the target
         let latest_synced_ledger_info =
             utils::fetch_latest_synced_ledger_info(self.storage.clone())?;
         self.consensus_notification_handler
             .check_sync_request_progress(latest_synced_ledger_info)
-            .await
+            .await?;
+
+        // If the sync request was successfully handled, reset the continuous syncer
+        // so that in the event another sync request occurs, we have a fresh state.
+        if !self.active_sync_request() {
+            self.continuous_syncer.reset_active_stream();
```

Comment:
this feels a weird place to reset it, shouldn't we reset it when terminate the stream?

Reply:
In an ideal world, this is true. The last two notifications along the stream should be: (i) the last chunk taking us to the target version; and (ii) the end of stream notification. When the continuous syncer reads the end of stream notification, it terminates the stream and clears the state.

However, the problem comes when consensus sends another sync request. If we don't reset the stream here, the continuous syncer will pick up the old stream and try to process any data chunks. If a malicious peer managed to send more data than requested, it will try to process those chunks. Those data items should be invalid, and so the stream will be terminated anyway, but, I'm just being very defensive here and clearing the stream now to avoid us ever getting into that state. Does that make sense?